### PR TITLE
feat(nm): Added Network Configuration enforcement

### PR DESCRIPTION
**Pull request by salvatore-coppola**
Added NMConnectionChangedHandler managing Network Configuration rollback.

Tests will be added in the pr #4408 once the pr will be merged.

### CODE REVIEW

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMConnectionChangedHandler.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    }
+
+    public Map<String, Object> getConfiguation() {
+        return configuation;
+    }
+
+    public void setConfiguation(Map<String, Object> configuation) {
+        this.configuation = configuation;
+    }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMConnectionChangedHandler.java]: Watch out! Typos... 

```suggestion
    public Map<String, Object> getConfiguration() {
        return configuration;
    }

    public void setConfiguration(Map<String, Object> configuration) {
        this.configuration = configuration;
    }
```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMConnectionChangedHandler.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.freedesktop.networkmanager.settings.Connection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("rawtypes")
+public class NMConnectionChangedHandler implements DBusSigHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(NMConnectionChangedHandler.class);
+
+    private Map<String, Object> configuation;
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMConnectionChangedHandler.java]: Why are you duplicating the cached configuration? It's already cached here: https://github.com/eclipse/kura/blob/4074e6e4c4809b48117352aabad7f7db32413505/kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java#L76

This adds the risk of using an old configuration.

You just need to call the `nmDbusConnector.apply()` method, which takes no argument and uses the cached configuration, like we do in the DeviceAdded signal handler

**Review by mattdibi — CHANGES_REQUESTED** [review]
Looks good apart from the configuration cache duplication and the typos. See my comments above.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMConnectionChangedHandler.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public void handle(DBusSignal signal) {
+        if (signal instanceof Settings.NewConnection || signal instanceof Settings.ConnectionRemoved
+                || signal instanceof Connection.Updated) {
+
+            logger.info("Detected external network change, rollbacking to the cached configuration!");
+            try {
+                this.nmDbusConnector.apply();
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMConnectionChangedHandler.java]: Is this still needed or could we just rely on the `getInstance()` method for retrieving a reference to the `nmDbusConnector`?

```suggestion
                NMDbusConnector nmDbusConnector = NMDbusConnector.getInstance(); 
                nmDbusConnector.apply();
```

**Review by mattdibi — CHANGES_REQUESTED** [review]
It is working as expected for the Connection Removed and Added signals.

It's not working properly for a settings update of an existing connection.

When I edit the `kura-wlan0-connection` managed by Kura using `nmtui` I get the following error in the Kura log:

```
2023-02-27T16:30:48,818 [qtp863784918-1077] INFO  o.e.j.s.h.C.ROOT - org.eclipse.kura.web.server.GwtComponentServiceImpl: Trying /www/denali/58535A5C98F1E514899ADE2185A43E60.gwt.rpc
2023-02-27T16:31:14,494 [DBus Worker Thread-2] WARN  o.f.d.m.DBusSignal - Could not find suitable constructor for class org.freedesktop.networkmanager.settings.Connection$Updated with argument-types: []
2023-02-27T16:31:40,323 [DBus Worker Thread-3] WARN  o.f.d.m.DBusSignal - Could not find suitable constructor for class org.freedesktop.networkmanager.settings.Connection$Updated with argument-types: []
2023-02-27T16:31:53,125 [DBus Worker Thread-4] WARN  o.f.d.m.DBusSignal - Could not find suitable constructor for class org.freedesktop.networkmanager.settings.Connection$Updated with argument-types: []
```

and no configuration roll-back is triggered.

Looks like the signal is thrown but the code fails to call the constructor correctly. There might be some thing missing in the `Connection.java` code.

**Comment by salvatore-coppola** [comment]
> It is working as expected for the Connection Removed and Added signals.
> 
> It's not working properly for a settings update of an existing connection.
> 
> When I edit the `kura-wlan0-connection` managed by Kura using `nmtui` I get the following error in the Kura log:
> 
> ```
> 2023-02-27T16:30:48,818 [qtp863784918-1077] INFO  o.e.j.s.h.C.ROOT - org.eclipse.kura.web.server.GwtComponentServiceImpl: Trying /www/denali/58535A5C98F1E514899ADE2185A43E60.gwt.rpc
> 2023-02-27T16:31:14,494 [DBus Worker Thread-2] WARN  o.f.d.m.DBusSignal - Could not find suitable constructor for class org.freedesktop.networkmanager.settings.Connection$Updated with argument-types: []
> 2023-02-27T16:31:40,323 [DBus Worker Thread-3] WARN  o.f.d.m.DBusSignal - Could not find suitable constructor for class org.freedesktop.networkmanager.settings.Connection$Updated with argument-types: []
> 2023-02-27T16:31:53,125 [DBus Worker Thread-4] WARN  o.f.d.m.DBusSignal - Could not find suitable constructor for class org.freedesktop.networkmanager.settings.Connection$Updated with argument-types: []
> ```
> 
> and no configuration roll-back is triggered.
> 
> Looks like the signal is thrown but the code fails to call the constructor correctly. There might be some thing missing in the `Connection.java` code.

i'll check

**Comment by mattdibi** [comment]
@salvatore-coppola Just tested it... something went terribly wrong: it is stuck in an infinite loop of configuration apply.

https://user-images.githubusercontent.com/22748355/222078786-e493a946-593f-4501-9d09-e48a614be654.mov

Looks like it is not ignoring the signals generated by Kura itself.

**Comment by salvatore-coppola** [comment]
> @salvatore-coppola Just tested it... something went terribly wrong: it is stuck in an infinite loop of configuration apply.
> 
> Looks like it is not ignoring the signals generated by Kura itself.

Wow, this is scary, i didn't notice that, need to investigate!

**Comment by mattdibi** [comment]
Superseded by https://github.com/eclipse/kura/pull/4434

**Pull request closed without merging**